Add configuration api http reference (#2569)
dapr/docs: daprdocs/content/en/reference/api/configuration_api.md

```json
{
  "id": "<unique-id>"
}
```

The returned `id` parameter can be used to unsubscribe to the specific set of keys provided on the subscribe API call. This should be retained by the application.

## Unsubscribe Configuration

This endpoint lets you unsubscribe to configuration changes.

### HTTP Request

```
GET http://localhost:<daprPort>/v1.0-alpha1/configuration/<subscription-id>/unsubscribe
```

#### URL Parameters

Parameter | Description
--------- | -----------
`daprPort` | The Dapr port
`subscription-id` | The value from the `id` field returned from the response of the subscribe endpoint

#### Query Parameters

None

#### Request Body

None

### HTTP Response

#### Response Codes

Code | Description
---- | -----------
`204`  | Unsubscribe operation successful
`400`  | Configuration store is missing or misconfigured or malformed request
`500`  | Failed to unsubscribe to configuration changes

#### Response Body

None

### Example

```shell
curl -X GET 'http://localhost:3500/v1.0-alpha1/configuration/bf3aa454-312d-403c-af95-6dec65058fa2/unsubscribe' 
```

## Optional application (user code) routes

### Provide a route for Dapr to send configuration changes

> When subscribing to configuration changes, Dapr invokes the application whenever a configuration item changes.

Reviewer: Your application can have a `/configuration` endpoint that is called for all key updates that are subscribed to. The endpoint(s) can be made more specific for a given configuration store by adding `/<store-name>` and for a specific key by adding  `/<store-name>/<key>` to the route.

> When subscribing to configuration changes, Dapr invokes the application whenever a configuration item changes.

Reviewer: ```suggestion
When subscribing to configuration changes, Dapr invokes the application whenever a configuration item changes. Your application can have a `/configuration` endpoint that is called for all key updates that are subscribed to. The endpoint(s) can be made more specific for a given configuration store by adding /<store-name> and for a specific key by adding /<store-name>/<key> to the route.
```

#### HTTP Request

```
POST http://localhost:<appPort>/configuration/<store-name>/<key>
```

#### URL Parameters

Parameter | Description
--------- | -----------
`appPort` | The application port

> `storename` | The `metadata.name` field component file. Refer to the [component schema] ({{< ref component-schema.md>}})

Reviewer: ```suggestion
`storename` | The `metadata.name` field component file. Refer to the [component schema] ({{< ref component-schema.md>}})
`key` | The key subscribed to
```

#### Request Body

> 

Reviewer: ```suggestion
A list of configuration items for a given subscription id. Configuration items can have a version associated with them, which is returned in the notification.
```

---
type: docs
title: "Configuration API reference"
linkTitle: "Configuration API"
description: "Detailed documentation on the configuration API"
weight: 650
---

## Get Configuration

This endpoint lets you get configuration from a store.

### HTTP Request

```
GET http://localhost:<daprPort>/v1.0-alpha1/configuration/<storename>
```

#### URL Parameters

Parameter | Description
--------- | -----------
`daprPort` | The Dapr port

> `storename` | The `metadata.name` field in the user-configured `configuration.yaml` component file. Refer to the [Dapr configuration store component file structure](#component-file) mentioned above.

Reviewer: ```suggestion
`storename` | The `metadata.name` field component file. Refer to the [component schema] ({{< ref component-schema.md>}})
```

Author: Changed

```

To retrieve all configuration items:

```
GET http://localhost:<daprPort>/v1.0-alpha1/configuration/mystore
```

#### Request Body

None

### HTTP Response

#### Response Codes

Code | Description
---- | -----------
`204`  | Get operation successful
`400`  | Configuration store is missing or misconfigured or malformed request
`500`  | Failed to get configuration

#### Response Body

JSON-encoded value of key/value pairs for each configuration item.

### Example

```shell
curl -X GET 'http://localhost:3500/v1.0-alpha1/configuration/mystore?key=myConfigKey' 
```

> The above command returns the following JSON:

```json
[{"key":"myConfigKey","value":"myConfigValue"}]
```

## Subscribe Configuration

This endpoint lets you subscribe to configuration changes. Notifications happen when values are updated or deleted in the configuration store. This enables the application to react to configuration changes.

### HTTP Request

```
GET http://localhost:<daprPort>/v1.0-alpha1/configuration/<storename>/subscribe
```

#### URL Parameters

Parameter | Description
--------- | -----------
`daprPort` | The Dapr port

> `storename` | The `metadata.name` field in the user-configured `configuration.yaml` component file. Refer to the [Dapr configuration store component file structure](#component-file) mentioned above.

Reviewer: ```suggestion
`storename` | The `metadata.name` field component file. Refer to the [component schema] ({{< ref component-schema.md>}})
```

Author: Changed

### HTTP Request

```
GET http://localhost:<daprPort>/v1.0-alpha1/configuration/<subscription-id>/unsubscribe
```

#### URL Parameters

Parameter | Description
--------- | -----------
`daprPort` | The Dapr port
`subscription-id` | The value from the `id` field returned from the response of the subscribe endpoint

#### Query Parameters

None

#### Request Body

None

### HTTP Response

#### Response Codes

Code | Description
---- | -----------
`204`  | Unsubscribe operation successful
`400`  | Configuration store is missing or misconfigured or malformed request
`500`  | Failed to unsubscribe to configuration changes

#### Response Body

None

### Example

```shell
curl -X GET 'http://localhost:3500/v1.0-alpha1/configuration/bf3aa454-312d-403c-af95-6dec65058fa2/unsubscribe' 
```

## Optional application (user code) routes

### Provide a route for Dapr to send configuration changes

When subscribing to configuration changes, Dapr invokes the application whenever a configuration item changes.

#### HTTP Request

```
POST http://localhost:<appPort>/configuration/<store-name>/<key>
```

#### URL Parameters

Parameter | Description
--------- | -----------
`appPort` | The application port

> `storename` | The `metadata.name` field in the user-configured `configuration.yaml` component file. Refer to the [Dapr configuration store component file structure](#component-file) mentioned above

Reviewer: ```suggestion
`storename` | The `metadata.name` field component file. Refer to the [component schema] ({{< ref component-schema.md>}})
```

Author: Changed

`key` | The key of the changed configuration item

#### Request Body

```json
{
    "id": "<subscription-id>",
    "items": [
        "key": "<key-of-configuration-item>",
        "value": "<new-value>",
        "version": "<version-of-item>"
    ]
}
```

#### Example

```json
{
    "id": "bf3aa454-312d-403c-af95-6dec65058fa2",
    "items": [
        "key": "config-1",
        "value": "abcdefgh",
        "version": "1.1"
    ]
}
```

## Next Steps

> - [Configuration api overview]({{< ref configuration-api-overview.md >}})

Reviewer: ```suggestion
- [Configuration API overview]({{< ref configuration-api-overview.md >}})
```

Author: Changed

> - [How-To: Manage configuration]({{< ref howto-manage-configuration.md >}})

Reviewer: ```suggestion
- [How-To: Manage configuration from a store]({{< ref howto-manage-configuration.md >}})
```

Author: Changed

---
type: docs
title: "Configuration API reference"
linkTitle: "Configuration API"
description: "Detailed documentation on the configuration API"

> weight: 700

Reviewer: ```suggestion
weight: 650
```

---

> ## Component file

Reviewer: You can remove this component file section. It is not needed here in the API spec.

> ## Component file

Author: Done

A Dapr `configuration.yaml` component file has the following structure:

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: <NAME>
  namespace: <NAMESPACE>
spec:
  type: configuration.<TYPE>
  version: v1
  metadata:
  - name:<KEY>
    value:<VALUE>
```

| Setting | Description                                                                            |
| ------- |----------------------------------------------------------------------------------------|
| `metadata.name` | The name of the configuration store.                                                   |

> | `spec.metadata` | An open key value pair metadata that allows a binding to define connection properties. |

Reviewer: This is a strange description
```suggestion
| `spec.metadata` | key/value pair metadata that is configured for each configuration store type. |
```

Author: This is not relevant anymore as this section was removed.

## Get Configuration

This endpoint lets you get configuration from a store.

### HTTP Request

> 

Reviewer: ```suggestion
Return all configuration items from the named store.
```

```
GET http://localhost:<daprPort>/v1.0/configuration/<storename>
```

#### URL Parameters

Parameter | Description
--------- | -----------
`daprPort` | The Dapr port
`storename` | The `metadata.name` field in the user-configured `configuration.yaml` component file. Refer to the [Dapr configuration store component file structure](#component-file) mentioned above.

#### Query Parameters

> If no query parameters are provided, all configuration items will be returned.

Reviewer: ```suggestion
If no query parameters are provided, all configuration items are returned.
```

> If no query parameters are provided, all configuration items will be returned.

Reviewer: Then Add this example to show retrieving all the items
```
GET http://localhost:<daprPort>/v1.0-alpha1/configuration/mystore
```

Author: Added

To specifiy the keys of the configuration items to get, use one or more `key` query parameters. For example:

```

> GET http://localhost:3500/v1.0/configuration/mystore?key=config1&key=config2

Reviewer: ```suggestion
GET http://localhost:3500/v1.0-alpha1/configuration/mystore?key=config1&key=config2
```

Author: Changed

```

#### Request Body

None

### HTTP Response

#### Response Codes

Code | Description
---- | -----------
`200`  | Get operation successful
`400`  | Configuration store is missing or misconfigured or malformed request
`500`  | Failed to get configuration

#### Response Body

> JSON-encoded value

Reviewer: ```suggestion
JSON-encoded value of key/value pairs for each configuration item.
```

Author: Changed

### Example

```shell

> curl -X GET 'http://localhost:3500/v1.0-alpha1/configuration/example-config?key=myConfigKey' 

Reviewer: ```suggestion
curl -X GET 'http://localhost:3500/v1.0-alpha1/configuration/mystore?key=myConfigKey' 
```

Author: Changed

```

> The above command returns the following JSON:

```json
[{"key":"myConfigKey","value":"myConfigValue"}]
```

## Subscribe Configuration

> This endpoint lets you subscribe to configuration changes.

Reviewer: ```suggestion
This endpoint lets you subscribe to configuration changes. Notifications happen when values are updated or deleted in the configuration store. This enables the application to react to configuration changes.
```

Author: Added

### HTTP Request

```

> GET http://localhost:<daprPort>/v1.0/configuration/<storename>/subscribe

Reviewer: ```suggestion
GET http://localhost:<daprPort>/v1.0-alpha1/configuration/<storename>/subscribe
```

Author: Changed

```

#### URL Parameters

Parameter | Description
--------- | -----------
`daprPort` | The Dapr port
`storename` | The `metadata.name` field in the user-configured `configuration.yaml` component file. Refer to the [Dapr configuration store component file structure](#component-file) mentioned above.

#### Query Parameters

> If no query parameters are provided, all configuration items will subscribed to.

Reviewer: ```suggestion
If no query parameters are provided, all configuration items are subscribed to.
```

Author: Changed

> If no query parameters are provided, all configuration items will subscribed to.

Reviewer: Then Add this example to show subscribing to all configuration item updates
```
GET http://localhost:<daprPort>/v1.0-alpha1/configuration/mystore/subscribe
```

Author: Added

To specifiy the keys of the configuration items to subscribe to, use one or more `key` query parameters. For example:

```

> GET http://localhost:3500/v1.0/configuration/mystore/subscribe?key=config1&key=config2

Reviewer: ```suggestion
GET http://localhost:3500/v1.0-alpha1/configuration/mystore/subscribe?key=config1&key=config2
```

Author: Changed

```

#### Request Body

None

### HTTP Response

#### Response Codes

Code | Description
---- | -----------
`200`  | Subscribe operation successful
`400`  | Configuration store is missing or misconfigured or malformed request
`500`  | Failed to subscribe to configuration changes

#### Response Body

JSON-encoded value

### Example

```shell

> curl -X GET 'http://localhost:3500/v1.0-alpha1/configuration/subscribe?key=myConfigKey' 

Reviewer: ```suggestion
curl -X GET 'http://localhost:3500/v1.0-alpha1/configuration/mystore/subscribe?key=myConfigKey' 
```

Author: Changed

```

> The above command returns the following JSON:

```json
{
  "id": "<unique-id>"
}
```

> The returned `id` parameter can be used to unsubscribe to the keys provided on the API call.

Reviewer: ```suggestion
The returned `id` parameter can be used to unsubscribe to the specific set of keys provided on the subscribe API call. This should be retained by the application.
```

Author: Changed

## Unsubscribe Configuration

This endpoint lets you unsubscribe to configuration changes.

### HTTP Request

```

> GET http://localhost:<daprPort>/v1.0/configuration/<subscription-id>/unsubscribe

Reviewer: Do you need store name here?
```suggestion
GET http://localhost:<daprPort>/v1.0-alpha1/configuration/<storename>/<subscription-id>/unsubscribe
```

Author: No

```

#### URL Parameters

Parameter | Description
--------- | -----------
`daprPort` | The Dapr port
`subscription-id` | The value from the `id` field returned from the response of the subscribe endpoint

#### Query Parameters

None

#### Request Body

None

### HTTP Response

#### Response Codes

Code | Description
---- | -----------
`200`  | Unsubscribe operation successful
`400`  | Configuration store is missing or misconfigured or malformed request
`500`  | Failed to unsubscribe to configuration changes

#### Response Body

None

### Example

```shell

> curl -X GET 'http://localhost:3500/v1.0-alpha1/configuration/bf3aa454-312d-403c-af95-6dec65058fa2/unsubscribe' 

Reviewer: Do you need store name here
```suggestion
curl -X GET 'http://localhost:3500/v1.0-alpha1/configuration/mystore/bf3aa454-312d-403c-af95-6dec65058fa2/unsubscribe' 
```

Author: No

```

> ## Optional Application (User Code) Routes

Reviewer: ```suggestion
## Optional application (user code) routes
```

Author: Changed

### Provide a route for Dapr to send configuration changes

> When subscribing to configuration changes, Dapr will invoke the application whenever a configuration item changes.

Reviewer: ```suggestion
When subscribing to configuration changes, Dapr can invoke an endpoint on the application whenever a configuration item changes.
```

Author: Changed

#### HTTP Request

```
POST http://localhost:<appPort>/configuration/<store-name>/<key>
```

#### URL Parameters

Parameter | Description
--------- | -----------
`appPort` | The application port
`storename` | The `metadata.name` field in the user-configured `configuration.yaml` component file. Refer to the [Dapr configuration store component file structure](#component-file) mentioned above
`key` | The key of the changed configuration item

> #### Request Body

Reviewer: What is the example showing here? is this supposed to show the routes on the application?

Author: Yes, we do the exact same thing on the pub/sub API reference.

Reviewer: If you are referring to this https://docs.dapr.io/reference/api/pubsub_api/#http-response-body then that is not what is shown below. What is `version`. And I do not understand how is this used? Whereas subscription for pub/sub returns a list of the routes, from the app, what is this calling on the app? Please explain.

Author: No, it's referring to https://docs.dapr.io/reference/api/pubsub_api/#provide-routes-for-dapr-to-deliver-topic-events.

Reviewer: Question: Like pub/sub are there any success/failure codes that the application can return to Dapr?

Author: No
